src/hotspot/cpu/x86/vm_version_x86.cpp

```diff
@@ -636,12 +1252,50 @@ void VM_Version::get_processor_features() {
 
   if (cpu_family() > 4) { // it supports CPUID
     _features = feature_flags();
+    _glibc_features = glibc_flags();
     // Logical processors are only available on P4s and above,
     // and only if hyperthreading is available.
     _logical_processors_per_package = logical_processor_count();
     _L1_data_cache_line_size = L1_line_size();
   }
 
+  if (ShowCPUFeatures) {
+    tty->print_cr("This machine's CPU features are: -XX:CPUFeatures=0x" UINT64_FORMAT_X ",0x" UINT64_FORMAT_X, _features, _glibc_features);
+  }
+
+  assert(!CPUFeatures == FLAG_IS_DEFAULT(CPUFeatures), "CPUFeatures parsing");
+  if (CPUFeatures) {
+    uint64_t GLIBCFeatures_x64;
+    uint64_t   CPUFeatures_x64 = CPUFeatures_parse(CPUFeatures, GLIBCFeatures_x64);
+    uint64_t       features_missing =   CPUFeatures_x64 & ~      _features;
+    uint64_t glibc_features_missing = GLIBCFeatures_x64 & ~_glibc_features;
+    if (features_missing || glibc_features_missing) {
+      char buf[512];
+      int res = jio_snprintf(
+                  buf, sizeof(buf),
+                  "Specified -XX:CPUFeatures=0x" UINT64_FORMAT_X ",0x" UINT64_FORMAT_X
+                  "; this machine's CPU features are 0x" UINT64_FORMAT_X ",0x" UINT64_FORMAT_X
+                  "; missing features of this CPU are 0x" UINT64_FORMAT_X ",0x" UINT64_FORMAT_X " ",
+                  CPUFeatures_x64, GLIBCFeatures_x64,
+                  _features, _glibc_features,
+                  features_missing, glibc_features_missing);
+      assert(res > 0, "not enough temporary space allocated");
+      // insert_features_names() does crash for undefined too high-numbered features.
+      insert_features_names(buf + res , sizeof(buf) - res ,       features_missing & (  CPU_MAX - 1));
+      int res2 = res + strlen(buf + res);
+      insert_features_names(buf + res2, sizeof(buf) - res2, glibc_features_missing & (GLIBC_MAX - 1));
+      if (buf[res]) {
+        assert(buf[res] == ',', "unexpeced VM_Version::insert_features_names separator instead of ','");
+        buf[res] = '=';
+      }
+      vm_exit_during_initialization(buf);
+    }
+          _features =   CPUFeatures_x64;
+    _glibc_features = GLIBCFeatures_x64;
+  }
+
+  glibc_not_using((CPU_MAX - 1) & ~_features, (GLIBC_MAX - 1) & ~_glibc_features);
```

Comment:
Apparently this triggers re-exec even with default command line parameters, which does not look expected, see below.

Also, we need a way to disable CPU feature reduction as a workaround to possible bugs in the implementation.

```
$ ./build/linux-x86_64-server-release/images/jdk/bin/java -XX:+ShowCPUFeatures -version
This machine's CPU features are: -XX:CPUFeatures=0x61805fdfbff,0x1e6
Re-exec of java with new environment variable: GLIBC_TUNABLES=:glibc.cpu.hwcaps=,-RTM,-AVX512F,-AVX512CD,-AVX512BW,-AVX512DQ,-AVX512ER,-AVX512PF,-AVX512VL,-IBT,-FMA4,-SHSTK
This machine's CPU features are: -XX:CPUFeatures=0x61805fdfbff,0x1e6
Environment variable already set, both glibc CPU_FEATURE_ACTIVE and ld.so --list-diagnostics are unavailable - re-exec suppressed: GLIBC_TUNABLES=:glibc.cpu.hwcaps=,-RTM,-AVX512F,-AVX512CD,-AVX512BW,-AVX512DQ,-AVX512ER,-AVX512PF,-AVX512VL,-IBT,-FMA4,-SHSTK
openjdk version "17-internal" 2021-09-14
OpenJDK Runtime Environment (build 17-internal+0-adhoc..crac)
OpenJDK 64-Bit Server VM (build 17-internal+0-adhoc..crac, mixed mode, sharing)
```

Reply:
I have remembered that was intentional for more safety of what jdk vs. glibc detect from CPU. I agree it is too expensive to do a jdk re-exec each time. The behavior you have seen can be now achieved by `-XX:CPUFeatures=verify`. Usefulness of this new option may not be too big.

To disable everything around the CPU features is now possible by `-XX:CPUFeatures=ignore`. I agree it may be useful if the glibc interface for these CPU features gets somehow incompatible.

Reply:
> I have remembered that was intentional for more safety of what jdk vs. glibc detect from CPU.

Could you elaborate a bit about the safety? In JDK we should have the same algorithm for determing glibc features. Do you mean the situation when JDK and GLIBC feature detection have diverged, e.g. due to update in the glibc?

Reply:
I have removed the option `-XX:CPUFeatures=verify`. It had some effect only on old glibcs and even there it did only check jdk against jdk's own logic. Not against glibc.

Diversion of JDK and GLIBC feature detection is tested by default but only if either `CPU_FEATURE_ACTIVE` or `ld.so --list-diagnostics` are available